runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -897,33 +949,93 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
             }
          }
 
-      // If the array is known to be of a primitive value type that is not flattened,
-      // transform the helper call to regular aaload and aastore.
-      bool canTransformUnflattenedVTArrayElementLoadStore = false;
-      if (arrayConstraint &&
+      // Transform the helper call to regular aaload and aastore if array flattening is not enabled,
+      // or the array is known to be of a primitive value type that is not flattened.
+      bool canTransformUnflattenedArrayElementLoadStore = TR::Compiler->om.isValueTypeArrayFlatteningEnabled() ? false : true;
+      if (!canTransformUnflattenedArrayElementLoadStore &&
+          arrayConstraint &&
           (isCompTypePrimVT == TR_yes) &&
-          (TR::Compiler->cls.isValueTypeClassFlattened(arrayConstraint->getClass()) == TR_no))
+          !TR::Compiler->cls.isValueTypeClassFlattened(arrayConstraint->getClass()))
          {
-         canTransformUnflattenedVTArrayElementLoadStore = true;
+         canTransformUnflattenedArrayElementLoadStore = true;
          }
 
       // If the array is known to have a component type that is not a primitive value type or
       // the value being stored is known not to be a value type, transform the helper
       // call to a regular aaload or aastore
-      bool canTransformIdentityLoadStore = false;
+      bool canTransformIdentityArrayElementLoadStore = false;
       if ((arrayConstraint != NULL && isCompTypePrimVT == TR_no)
           || (isStoreFlattenableArrayElement && isStoreValueVT == TR_no))
          {
-         canTransformIdentityLoadStore = true;
+         canTransformIdentityArrayElementLoadStore = true;
+         }
+
+      bool canTransformFlattenedArrayElementLoadStoreUseTypeHint = false;
+      bool canTransformUnflattenedArrayElementLoadStoreUseTypeHint = false;
+      bool canTransformIdentityArrayElementLoadStoreUseTypeHint = false;
+      static const char *disableFlattenedArrayElementTypeHintXForm = feGetEnv("TR_DisableFlattenedArrayElementTypeHintXForm");
+      static const char *disableUnflattenedArrayElementTypeHintXForm = feGetEnv("TR_DisableUnflattenedArrayElementTypeHintXForm");
+      TR_OpaqueClassBlock *typeHintClass = arrayConstraint ? arrayConstraint->getTypeHintClass() : NULL;
+
+      if (typeHintClass &&
+          !canTransformFlattenedArrayElementLoadStore &&
+          !canTransformUnflattenedArrayElementLoadStore &&
+          !canTransformIdentityArrayElementLoadStore)
+         {
+         TR_OpaqueClassBlock *componentClass = comp()->fej9()->getComponentClassFromArrayClass(typeHintClass);
+
+         if (!disableUnflattenedArrayElementTypeHintXForm &&
+             TR::Compiler->cls.classHasIdentity(componentClass))
```

Comment:
It just occurred to me that there is already something similar for the identity case in `TreeLowering.cpp`.  If the type hint is for an identity type, we're going to end up with IL that checks whether the array component type is a particular identity type, and if it's not, whether it is a primitive value type.

You might want to do some performance measurements to see how performance looks if the type hint is an identity class and the array component type matches, versus the case where the type hint is an identity class and the array component type does not match, both with and without `TR_DisableUnflattenedArrayElementTypeHintXForm` set.

I don't think this comment needs to hold up this change.